src/plugins/visualizations/public/embeddable/visualize_embeddable.ts

```diff
@@ -381,4 +412,52 @@ export class VisualizeEmbeddable extends Embeddable<VisualizeInput, VisualizeOut
   public supportedTriggers(): TriggerId[] {
     return this.vis.type.getSupportedTriggers?.() ?? [];
   }
+
+  inputIsRefType = (input: VisualizeInput): input is VisualizeByReferenceInput => {
+    if (!this.attributeService) {
+      throw new Error('AttributeService must be defined for getInputAsRefType');
+    }
+    return this.attributeService.inputIsRefType(input as VisualizeByReferenceInput);
+  };
+
+  getInputAsValueType = async (): Promise<VisualizeByValueInput> => {
```

Comment:
Why not use the attribute service here as well? The implementation is quite simple, but it could be more consistent that way.

Reply:
I tried, but it's honestly too much hassle. I'd need a `customUnwrapMethod` and the code becomes so much more complicated. I don't think it's worth the effort.